src/librustc_lint/non_ascii_idents.rs

```diff
@@ -17,157 +15,237 @@ declare_lint! {
+                        continue;
+                    }
+                    if !GeneralSecurityProfile::identifier_allowed(ch) {
+                        // this character is covered by `uncommon_codepoints` lint.
+                        continue;
+                    }
+                    let augmented_script_set = AugmentedScriptSet::for_char(ch);
+                    script_states
+                        .entry(augmented_script_set)
+                        .and_modify(|existing_state| match existing_state {
+                            ScriptSetUsage::Verified => {}
+                            ScriptSetUsage::Suspicious(_, _) => {
+                                if !is_potential_mixed_script_confusable_char(ch) {
+                                    *existing_state = ScriptSetUsage::Verified;
+                                }
+                            }
+                        })
+                        .or_insert_with(|| {
+                            if !is_potential_mixed_script_confusable_char(ch) {
+                                ScriptSetUsage::Verified
+                            } else {
+                                ScriptSetUsage::Suspicious(ch, sp)
+                            }
+                        });
+                }
+            }
+
+            let verified_augmented_script_sets = script_states
+                .iter()
+                .flat_map(|(k, v)| match v {
+                    ScriptSetUsage::Verified => Some(*k),
+                    _ => None,
                 })
-                .or_insert((symbol_str, sp));
+                .collect::<Vec<_>>();
+
+            for (augment_script_set, usage) in script_states {
+                let (ch, sp) = match usage {
+                    ScriptSetUsage::Verified => continue,
+                    ScriptSetUsage::Suspicious(ch, sp) => (ch, sp),
+                };
+
+                if augment_script_set.is_all() {
+                    continue;
+                }
+
+                let mut has_intersect = false;
+                'find_intersect: for existing in verified_augmented_script_sets.iter() {
+                    if existing.is_all() {
+                        continue;
+                    }
+                    let mut intersect = *existing;
+                    intersect.intersect_with(augment_script_set);
+                    if !intersect.is_empty() && !intersect.is_all() {
+                        has_intersect = true;
+                        break 'find_intersect;
+                    }
+                }
+
+                if has_intersect {
+                    continue;
+                }
+
+                cx.struct_span_lint(MIXED_SCRIPT_CONFUSABLES, sp, |lint| {
+                    let message = format!(
+                        "Unicode augmented script group `{:?}` usage in this crate consists solely of mixed script confusables, including {:?} and maybe other characters.",
```

Comment:
"The usage of Script Group `{:?}` in this crate consists solely of mixed script confusables, including `{:?}`"

Reply:
Done in principle. Didn't find out a proper way to generate a long error message, so i moved the second sub-sentence to a note. Please review again to see if this is ok.